File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/7b2ed196ac86d9c6a19962e13d67ff47.py
import sys
from collections import defaultdict
from itertools import accumulate

class DSum:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self._in = in_stream
            self._out = out_stream
            self.k = 0
            self.dps = []
            self.sums = []
            self.arrs = []

        def rd(self):
            return map(int, self._in.readline().split())

        def pr(self, *args):
            self._out.write(' '.join(map(str, args)))

        def prln(self, *args):
            self.pr(*args)
            self._out.write('\n')

        def do_copy(self, dpi):
            self.dps[dpi + 1] = self.dps[dpi][:]

        def add_to_knap(self, i, dpi):
            v = self.sums[i]
            k1 = len(self.arrs[i])
            for j in range(self.k, k1 - 1, -1):
                self.dps[dpi][j] = max(self.dps[dpi][j], self.dps[dpi][j - k1] + v)

        def rec(self, l, r, dpi):
            ans = 0
            carr = self.arrs[l]
            dp = self.dps[dpi]
            if l + 1 == r:
                ans = max(ans, dp[self.k])
                sum_val = 0
                for i in range(min(self.k, len(carr))):
                    sum_val += carr[i]
                    ans = max(ans, dp[self.k - i - 1] + sum_val)
            else:
                m = (l + r) // 2
                self.do_copy(dpi)
                for i in range(l, m):
                    self.add_to_knap(i, dpi + 1)
                ans = max(ans, self.rec(m, r, dpi + 1))
                self.do_copy(dpi)
                for i in range(m, r):
                    self.add_to_knap(i, dpi + 1)
                ans = max(ans, self.rec(l, m, dpi + 1))
            return ans

        def solve(self):
            n, self.k = self.rd()
            self.sums = [0] * n
            self.arrs = []
            for i in range(n):
                t, *arr = self.rd()
                self.arrs.append(arr[:self.k])
                self.sums[i] = sum(self.arrs[i])
            self.dps = [[0] * (self.k + 1) for _ in range(13)]
            self.prln(self.rec(0, n, 0))

    @staticmethod
    def solve(in_stream, out_stream):
        DSum.Solver(in_stream, out_stream).solve()

if __name__ == "__main__":
    input_stream = sys.stdin
    output_stream = sys.stdout
    DSum.solve(input_stream, output_stream)

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/C/7b2ed196ac86d9c6a19962e13d67ff47.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdbool.h>

#define MAX_N 100
#define MAX_K 100

long long sums[MAX_N];
long long dps[13][MAX_K + 1];
long long arrs[MAX_N][MAX_K];
int arr_sizes[MAX_N];

long long sum(int *arr, int size) {
    long long total = 0;
    for (int i = 0; i < size; i++) {
        total += arr[i];
    }
    return total;
}

void do_copy(int dpi, int k) {
    memcpy(dps[dpi + 1], dps[dpi], sizeof(dps[0][0]) * (k + 1));
}

void add_to_knap(int i, int dpi, int k) {
    long long v = sums[i];
    int k1 = arr_sizes[i];
    for (int j = k; j >= k1; j--) {
        if (dps[dpi][j] < dps[dpi][j - k1] + v) {
            dps[dpi][j] = dps[dpi][j - k1] + v;
        }
    }
}

long long rec(int l, int r, int dpi, int k) {
    long long ans = 0;
    int *carr = arrs[l];
    long long *dp = dps[dpi];

    if (l + 1 == r) {
        ans = dp[k];
        long long sum_val = 0;
        for (int i = 0; i < (k < arr_sizes[l] ? k : arr_sizes[l]); i++) {
            sum_val += carr[i];
            if (dp[k - i - 1] + sum_val > ans) {
                ans = dp[k - i - 1] + sum_val;
            }
        }
    } else {
        int m = (l + r) / 2;
        do_copy(dpi, k);
        for (int i = l; i < m; i++) {
            add_to_knap(i, dpi + 1, k);
        }
        long long temp_ans = rec(m, r, dpi + 1, k);
        if (temp_ans > ans) {
            ans = temp_ans;
        }
        do_copy(dpi, k);
        for (int i = m; i < r; i++) {
            add_to_knap(i, dpi + 1, k);
        }
        temp_ans = rec(l, m, dpi + 1, k);
        if (temp_ans > ans) {
            ans = temp_ans;
        }
    }
    return ans;
}

void solve() {
    int n, k;
    scanf("%d %d", &n, &k);
    for (int i = 0; i < n; i++) {
        int t;
        scanf("%d", &t);
        arr_sizes[i] = t;
        for (int j = 0; j < t; j++) {
            scanf("%lld", &arrs[i][j]);
        }
        if (t > k) {
            arr_sizes[i] = k;
        }
        sums[i] = sum(arrs[i], arr_sizes[i]);
    }
    memset(dps, 0, sizeof(dps));
    printf("%lld\n", rec(0, n, 0, k));
}

int main() {
    solve();
    return 0;
}
